### PR TITLE
Update harness to use integrated vdom from widget-core

### PR DESCRIPTION
**Pull request by kitsonk**
**Type:** feature

The following has been addressed in the PR:

* [ ] There is a related issue
* [X] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [X] Unit or Functional tests are included in the PR

**Description:**

This PR updates the harness to be compatible with the integrated vdom from widget-core (PR dojo/widget-core#728.

It also makes a few other changes, some of them breaking:

* Uses the Projector instead of accessing the virtual DOM projection directly.
* ~Uses the Projector in a sandbox mode, therefore widgets are rendered on a document fragment instead of actually in the DOM.  This is more consistent with the concepts of harnessing and unit testing.~ Actually that was causing issues on Safari I couldn't fully resolve.  Basically Safari 9 doesn't like dispatching events on DOM owned by a document fragment.
* Changes the types to make the harness easier to use with no need to explicitly pass generics with widgets.  Also `.setProperties` and `.setChildren` will now properly type guard.  This is a **breaking change** in that many cases people might have had to specify two generic arguments to have the harness work.

This will need to be updated once the upstream PR is merged and published.

### CODE REVIEW

**Comment by kitsonk** [comment]
The difference in code coverage is a red herring.  Because there was quite a bit of code stripped because of leveraging the projector mixin from widget core, it is causing the code coverage to be 😭

File: src/support/assertRender.ts [hunk 1]
```diff
@@ -66,40 +61,47 @@ export default function assertRender(actual: DNode | DNode[], expected: DNode |
...
-			assertRender(actualChild, expected[index], message);
-		});
 	}
 
 	if (Array.isArray(actual) && Array.isArray(expected)) {
 		assertChildren(actual, expected);
 	}
 	else if (Array.isArray(actual) || Array.isArray(expected)) {
-		throwAssertionError(actual, expected, message);
+		throwAssertionError(actual, expected, Array.isArray(actual) ? `Expected "${expected === null ? 'null' : typeof expected}" but got an array` : `Expected an array but got "${actual === null ? 'null' : typeof actual}"`, message);
```
Comment (agubler) [inline src/support/assertRender.ts]: nit: This is quite long

File: src/harness.ts [hunk 2]
```diff
@@ -500,9 +365,9 @@ export class Harness<W extends WidgetBase<WidgetProperties>> extends Evented {
 	public sendEvent<I extends EventInit>(type: string, options: HarnessSendEventOptions<I> = {}): this {
 		let { target = this.getDom(), key, ...sendOptions } = options;
 		if (key) {
-			const vnode = findVNodebyKey(this._currentRender, key);
-			if (vnode && vnode.domNode) {
-				target = vnode.domNode as Element;
+			const dnode = findDNodeByKey(this._widgetHarness.lastRender, key);
+			if (dnode && !isWNode(dnode)) {
```
Comment (agubler) [inline src/harness.ts]: any reason why not `isHNode` instead of `!isWNode`?

**Review by agubler — COMMENTED** [review]
Couple of small comments

**Comment by kitsonk** [comment]
The coverage is a bit of a red herring, as it is more related to the changes in widget-core causing some less branch coverage, plus the overall reduction of code because of the removal of the widget.classes logic.

File: src/harness.ts (hunk 2 above)
Comment (kitsonk) [reply src/harness.ts]: Not that I can think of... Wasn't thinking.

File: src/support/assertRender.ts (hunk 1 above)
Comment (kitsonk) [reply src/support/assertRender.ts]: I refactored all the messages, you are right it was messy and unreadable.

**Comment by kitsonk** [comment]
I accidentally committed some change to the diagnostics of assert render that I had in another branch which I was going to submit after this PR, but they leaked in, so I merged the rest of them.  Overall it just improves the information you get from assert render when something goes wrong, instead of scratching your head, so hopefully we can turn a blind eye.

**Review by agubler — APPROVED** [review]

**Pull request merged**